### DISCUSSION TITLE
Code completion for settings files

### DISCUSSION THREAD

**Discussion by flcdrg** [statiqdev/Discussions#211] [Ideas]
One observation with using YAML and JSON for settings is that you miss out on the tooling/code completion experience. I was wondering if it would be possible to have a settings file (maybe .chtml, .csx or .cs) that tooling would recognise and thus be able to give you code completion and syntax highlighting?

I know you can do settings in Program.cs but thinking also for the equivalent of _directory.yml

**Comment by OskarKlintrot** [comment]
Microsoft uses JSON schema a lot, that works great imho!

**Reply by flcdrg** [reply]
But you wouldn't get C# code completion when using a JSON file would you?

**Reply by OskarKlintrot** [reply]
Not sure I understand the question, do you need C# code completion or just code completion? This is what it looks like:

![image](https://github.com/statiqdev/Discussions/assets/6783879/152f2e0c-9cf0-4c62-9661-299aad68469d)

[Here's the actual schema being used](https://json.schemastore.org/template)

**Reply by flcdrg** [reply]
I'm wanting C# code completion (hence the suggestion about .cs/csx/cshtml for settings)

**Reply by OskarKlintrot** [reply]
And https://github.com/statiqdev/Statiq.Web/blob/main/src/Statiq.Web/WebKeys.cs doesn't suffice?

**Reply by flcdrg** [reply]
I may not have been clear in my original question. It's not so much the property names (left hand side names) that I'm interested in getting Intellisense for, but when you're typing an expression in on the right hand side.

eg. I'd like proper C# syntax highlighting and Intellisense/code completion available when I'm editing the code between the curly braces

```yaml
DestinationPath: |
  =>
  {
    var pages = new NormalizedPath("pages");
    var relativePath = pages.GetRelativePath(Document.Destination);
    return relativePath.ChangeExtension("html");
  }
```

**Comment by TheBrambleShark** [comment]
Microsoft has published documentation on using [XML Schema Definitions](https://learn.microsoft.com/en-us/visualstudio/xml-tools/how-to-create-an-xml-schema-from-an-xml-document?view=vs-2022) with intellisense. This is confirmed to work in Visual Studio and Visual Studio Code. Rider has a [similar](https://www.jetbrains.com/help/rider/Referencing_XML_Schemas_and_DTDs.html) document regarding using XML Schemas in that IDE, for those who prefer that.

If we're looking for something along these lines, XML may be the way to go. Most properties here will either be open-ended where the consumer defines their own custom keys and values, or in the case of built-ins, they'll either be booleans, integers, or an enum value. Specifying this in the schema will allow us to get similar intellisense behavior to C# XML documentation.
